src/main/java/com/hillel/items_exchange/service/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+
+    @Override
+    public CategoryDto update(CategoryDto categoryDto) {
+        final Category updatedCategory = saveCategory(categoryDto);
+        return convertTo(updatedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public void removeById(long categoryId) {
+        categoryRepository.deleteById(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoDeletable(long categoryId) {
+        return isCategoryExistsById(categoryId) && isInternalSubcategoriesHaveNotProducts(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoUpdatable(CategoryDto categoryDto) {
+        return isCategoryExistsByIdAndNameOrNotExistsByName(categoryDto.getId(),
+                categoryDto.getName()) && isSubcategoryHasIdZeroOrIdExists(categoryDto);
+    }
+
+    @Override
+    public boolean isCategoryDtoValidForCreating(CategoryDto categoryDto) {
+        return isCategoryNameHasNotDuplicate(categoryDto.getName())
+                && categoryDto.getSubcategories().stream().allMatch(subcategoryDto -> subcategoryDto.getId() == 0L);
+    }
+
+    private boolean isCategoryExistsById(long categoryId) {
+        return categoryRepository.existsById(categoryId);
+    }
+
+    private boolean isCategoryNameHasNotDuplicate(String name) {
+        return !categoryRepository.existsByNameIgnoreCase(name);
+    }
+
+    private boolean isSubcategoryHasIdZeroOrIdExists(CategoryDto categoryDto) {
+        return categoryDto.getSubcategories().stream()
+                .filter(subcategoryDto -> subcategoryDto.getId() != 0)
+                .allMatch(subcategoryDto -> subcategoryService.isSubcategoryExistsById(subcategoryDto.getId()));
+    }
+
+    private boolean isInternalSubcategoriesHaveNotProducts(long categoryId) {
+        return categoryRepository.findById(categoryId)
+                .map(category -> category.getSubcategories().stream()
+                        .allMatch(subcategory -> subcategory.getProducts().isEmpty()))
+                .orElse(false);
+    }
+
+    private boolean isCategoryExistsByIdAndNameOrNotExistsByName(long categoryId, String categoryName) {
+        return categoryRepository.existsByIdAndNameIgnoreCase(categoryId, categoryName)
+                || (categoryRepository.existsById(categoryId) && isCategoryNameHasNotDuplicate(categoryName));
+    }
+
+    private Category saveCategory(CategoryDto categoryDto) {
+        final Category category = convertTo(categoryDto, Category.class);
+        List<Subcategory> subcategories = mapBy(categoryDto.getSubcategories(),
+                subcategoryDto -> setCategoryToSubcategory(category, subcategoryDto));
+        category.setSubcategories(subcategories);
+        return categoryRepository.saveAndFlush(category);
+    }
+
+    private Subcategory setCategoryToSubcategory(Category category, SubcategoryDto subcategoryDto) {
+        final Subcategory subcategory = modelMapper.map(subcategoryDto, Subcategory.class);
+        subcategory.setCategory(category);
```

Comment:
Do we really need bedirectional connection?

Reply:
I couldn't do it any other way because I was having problems persisting entities in the DB